spinalcordtoolbox/qmri/flow.py

```diff
@@ -0,0 +1,77 @@
+"""
+This is the interface API to compute flow-related metrics
+
+Copyright (c) 2023 Polytechnique Montreal <www.neuro.polymtl.ca>
+Author: Julien Cohen-Adad & ChatGPTv4
+License: see the file LICENSE
+"""
+
+import logging
+
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+
+def calculate_velocity(data_phase, venc):
+    """
+    Convert phase data to velocity. Assumes phase data is scaled between -pi and pi.
+    :param data_phase: 4D numpy array of phase data. The 4th dimension should be the velocity encoding (VENC) in cm/s.
+    :param venc: Maximum velocity that can be encoded in cm/s.
+    :return: velocity data
+    """
+
+    # Check that phase data is scaled between -pi and pi
+    if np.any(data_phase > np.pi) or np.any(data_phase < -np.pi):
+        logger.warning('Phase data is not scaled between -pi and pi. Please check your data.')
+
+    # Phase to velocity conversion
+    velocity = np.true_divide(data_phase * venc, np.pi)
+    return velocity
+
+# def calculate_flow(velocity_data, pixel_area):
+#     """
+#     Calculate flow by integrating velocity over a given area.
+#     :param velocity_data: numpy array of velocity data
+#     :param pixel_area: area of each pixel in the image (in square meters)
+#     :return: flow rate in the vessel (in cubic meters per second)
+#     """
+#     flow_rate = np.sum(velocity_data) * pixel_area
+#     return flow_rate
+
+
+def compute_flow(nii_phase4d, venc):
+    """
+    Compute velocity from VENC technique. More details at: https://mriquestions.com/what-is-venc.html
+    :param nii_phase4d: 4D phase image. The 4th dimension should be the velocity encoding (VENC) in cm/s.
+    :param venc: float: Maximum velocity that can be encoded
+    :return: nii_flow
+    """
+    # Convert input to avoid numerical errors from int16 data
+    # Related issue: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3636
+    nii_phase4d.change_type('float32')
+
+    # Initialize Image object
+    nii_flow = nii_phase4d.copy()
+
+    # Compute flow
+    nii_flow.data = np.true_divide(nii_phase4d.data, 2 * np.pi * venc)
+
+    return nii_flow
+
+
+def scale_phase(data_phase):
+    """
+    Scale phase data between -pi and pi, assuming the data is encoded as int13 (i.e., between -4096 and 4095)
+    as this seems to be the default encoding for phase data from Siemens scanners.
+    :param data_phase: numpy array of phase data
+    :return: scaled phase data
+    """
+    # Check that phase data is encoded as int13
+    if np.any(data_phase > 4095) or np.any(data_phase < -4096):
+        logger.warning('Phase data is not encoded as INT13 (i.e., between -4096 and 4095). Please check your data.')
+    # Convert data to float
+    data_phase = data_phase.astype(np.float32)
+    # Scale the data from [-4096, 4095] to [-π, π]
+    scaled_data = (data_phase / 4096) * np.pi
+    return scaled_data
```

Comment:
To provide a more conclusive answer about the range of data, I found https://github.com/rordenlab/dcm2niix/issues/406:

> This sample follows the convention I have seen: the raw data is in the range 0..4095 (12-bit), with the [scl_slope = 2.0 and scl_inter = -4096](https://nifti.nimh.nih.gov/nifti-1/documentation/nifti1fields/nifti1fields_pages/scl_slopeinter.html) yielding values of -4096..4094. A common trick is to assume the full range is -pi..pi.

In other words, I think the current code might be slightly flawed:

- It mentions the range `[-4096, 4095]`, rather than the actual range of `[-4096, 4094]`
- It performs the scaling by dividing by `4096`, when I think we might need to perform a slightly more involved scaling operation (due to the mention of slope/inter == 2,-4096):

```suggestion
def scale_phase(data_phase):
    """
    Scale phase data between -pi and pi. We assume the data was originally encoded as int12 [0, 4095] (as this seems
    to be the default encoding for phase data from Siemens scanners), then scaled to [-4096, 4094] during conversion
    from DICOM to Nifti due to a slope/intercept of 2.0 and -4096.

    See also: https://github.com/rordenlab/dcm2niix/issues/406#issuecomment-654141193

    :param data_phase: numpy array of phase data
    :return: scaled phase data
    """
    # Check that phase data is encoded as scaled 12-bit integers
    slope = 2.0
    intercept = -4096
    bounds_unscaled = [0, 4095]
    bounds_scaled = [b * slope + intercept for b in bounds_unscaled]
    if np.any(data_phase < bounds_scaled[0]) or np.any(data_phase < bounds_scaled[1]):
        logger.warning(f'Phase data is not within the expected range for Siemens phase data (i.e., between '
                       f'{bounds_scaled[0]} and {bounds_scaled[1]}). Please check your data.')
    # Convert data to float
    data_phase = data_phase.astype(np.float32)
    # Scale the data from [-4096, 4094] back to [0, 4095]
    scaled_data = (data_phase - intercept) / slope
    # Scale the data from [0, 4095] to [0, 2]
    scaled_data = scaled_data / bounds_unscaled[1] * 2
    # Scale the data from [0, 2] to [-π, π]
    scaled_data = (scaled_data - 1) * np.pi
    return scaled_data
```

Reply:
Before: 

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/2a89ae15-44a2-459d-9928-84a593b5c4f1)

After:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/4f9f04bc-b762-4083-8938-65a49140b063)

Reply:
We looked over this revision and JCA gave the :+1: